lib/OnyxUtils.ts

```diff
@@ -1142,77 +1145,101 @@ function hasPendingMergeForKey(key: OnyxKey): boolean {
+        }
+    });
 
-    // We can remove all null values in an object by merging it with itself
-    // utils.fastMerge recursively goes through the object and removes all null values
-    // Passing two identical objects as source and target to fastMerge will not change it, but only remove the null values
-    return {value: shouldRemoveNestedNulls ? utils.removeNestedNullValues(value) : value, wasRemoved: false};
+    return pairs;
 }
 
 /**
- * Storage expects array like: [["@MyApp_user", value_1], ["@MyApp_key", value_2]]
- * This method transforms an object like {'@MyApp_user': myUserValue, '@MyApp_key': myKeyValue}
- * to an array of key-value pairs in the above format and removes key-value pairs that are being set to null
-
-* @return an array of key - value pairs <[key, value]>
+ * Merges an array of changes with an existing value or creates a single change.
+ *
+ * @param changes Array of changes that should be merged
+ * @param existingValue The existing value that should be merged with the changes
  */
-function prepareKeyValuePairsForStorage(data: Record<OnyxKey, OnyxInput<OnyxKey>>, shouldRemoveNestedNulls: boolean): Array<[OnyxKey, OnyxInput<OnyxKey>]> {
-    return Object.entries(data).reduce<Array<[OnyxKey, OnyxInput<OnyxKey>]>>((pairs, [key, value]) => {
-        const {value: valueAfterRemoving, wasRemoved} = removeNullValues(key, value, shouldRemoveNestedNulls);
-
-        if (!wasRemoved && valueAfterRemoving !== undefined) {
-            pairs.push([key, valueAfterRemoving]);
-        }
+function mergeChanges<TValue extends OnyxInput<OnyxKey> | undefined, TChange extends OnyxInput<OnyxKey> | undefined>(changes: TChange[], existingValue?: TValue): FastMergeResult<TChange> {
+    return applyMerge('merge', changes, existingValue);
+}
 
-        return pairs;
-    }, []);
+/**
+ * Merges an array of changes with an existing value or creates a single change.
+ * It will also mark deep nested objects that need to be entirely replaced during the merge.
+ *
+ * @param changes Array of changes that should be merged
+ * @param existingValue The existing value that should be merged with the changes
+ */
+function mergeAndMarkChanges<TValue extends OnyxInput<OnyxKey> | undefined, TChange extends OnyxInput<OnyxKey> | undefined>(
+    changes: TChange[],
+    existingValue?: TValue,
+): FastMergeResult<TChange> {
+    return applyMerge('mark', changes, existingValue);
 }
 
 /**
- * Merges an array of changes with an existing value
+ * Merges an array of changes with an existing value or creates a single change.
  *
- * @param changes Array of changes that should be applied to the existing value
+ * @param changes Array of changes that should be merged
+ * @param existingValue The existing value that should be merged with the changes
  */
 function applyMerge<TValue extends OnyxInput<OnyxKey> | undefined, TChange extends OnyxInput<OnyxKey> | undefined>(
```

Comment:
To avoid confusion, could we rename this (internal) function to sth different, since we also have `OnyxMerge.applyMerge`.

This was my mistake when i refactored this part but i think it would be good to have unique function names for further discussion. Maybe `mergeImplementation` or `mergeInternal` is appropriate, since we are not exposing this function anyway